compiler/rustc_mir_transform/src/cross_crate_inline.rs

```diff
@@ -50,6 +50,12 @@ fn cross_crate_inlinable(tcx: TyCtxt<'_>, def_id: LocalDefId) -> bool {
         _ => {}
     }
 
+    // If the crate is likely to be mostly unused, use cross-crate inlining to defer codegen until
+    // the function is referenced, in order to skip codegen for unused functions.
+    if tcx.sess.opts.cg.hint_mostly_unused {
+        return true;
+    }
+
```

Comment:
Placing this logic below the check for `InlineAttr::Never` does not change whether a call to a `#[inline(never)]` function can be inlined, it changes when and where `#[inline(never)]` functions get compiled. With the logic down here, `#[inline(never)]` functions and their callees (recursively) are not delayed.

The exact motivation for putting this logic here instead of above the check for `codegen_fn_attrs.inline` should be clearly recorded. I don't rightly remember what it was, but for example panic backtrace trimming still works even if it is moved.

Reply:
Making sure inline-never functions don't get affected was an intentional goal here. I'll add a comment.

Reply:
I've added a comment and documentation for this.